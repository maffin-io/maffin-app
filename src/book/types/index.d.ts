import type { DateTime } from "luxon";

export type QuoteInfo = {
  changePct: number,
  changeAbs: number,
  price: number,
  currency: string,
  date: DateTime,
};
