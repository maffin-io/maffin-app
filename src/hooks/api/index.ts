import { DateTime } from 'luxon';
import { mutate, SWRResponse } from 'swr';
import useSWRImmutable from 'swr/immutable';

import { Commodity } from '@/book/entities';
import { InvestmentAccount } from '@/book/models';
import * as queries from '@/lib/queries';
import type { PriceDBMap } from '@/book/prices';
import type { AccountsMap } from '@/types/book';
import type { Account, Split, Transaction } from '@/book/entities';

export function useStartDate(): SWRResponse<DateTime> {
  const key = '/api/start-date';
  return useSWRImmutable(
    key,
    fetcher(queries.getEarliestDate, key),
  );
}

export function useMainCurrency(): SWRResponse<Commodity> {
  const key = '/api/main-currency';
  return useSWRImmutable(
    key,
    fetcher(queries.getMainCurrency, key),
  );
}

export function useCommodity(guid: string): SWRResponse<Commodity> {
  const key = `/api/commodities/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.findOneByOrFail({ guid }), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useCommodities(): SWRResponse<Commodity[]> {
  const key = '/api/commodities';

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.find(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (commodity: Commodity) => mutate(
        `/api/commodities/${commodity.guid}`,
        commodity,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useLatestTxs(): SWRResponse<Transaction[]> {
  const key = '/api/txs/latest';
  return useSWRImmutable(
    key,
    fetcher(queries.getLatestTxs, key),
  );
}

export function useAccount(guid: string): SWRResponse<Account> {
  const key = `/api/accounts/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useAccounts(): SWRResponse<AccountsMap> {
  const key = '/api/accounts';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    Object.values(result.data as AccountsMap).forEach(
      (account: Account) => mutate(`/api/accounts/${account.guid}`, account, { revalidate: false }),
    );
  }

  return result;
}

export function useAccountsMonthlyTotals(): SWRResponse<queries.MonthlyTotals> {
  const { data: accounts } = useAccounts();
  const { data: prices } = useSWRImmutable(
    '/api/prices',
    fetcher(() => queries.getPrices({}), '/api/prices'),
  );

  const key = '/api/monthly-totals';
  const result = useSWRImmutable(
    (accounts && prices) ? key : null,
    fetcher(
      () => queries.getMonthlyTotals(accounts as AccountsMap, prices),
      key,
    ),
  );

  return result;
}

export function useInvestment(guid: string): SWRResponse<InvestmentAccount> {
  const key = `/api/investments/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestment(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useInvestments(): SWRResponse<InvestmentAccount[]> {
  const key = '/api/investments';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestments(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (investment: InvestmentAccount) => mutate(
        `/api/investments/${investment.account.guid}`,
        investment,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useSplits(guid: string): SWRResponse<Split[]> {
  const key = `/api/splits/${guid}`;
  return useSWRImmutable(
    key,
    fetcher(async () => queries.getSplits(guid), key),
  );
}

/**
 * Returns prices for a given commodity
 */
export function usePrices(c: Commodity | undefined): SWRResponse<PriceDBMap> {
  const key = `/api/prices/${c?.guid}`;
  return useSWRImmutable(
    c?.guid ? key : null,
    fetcher(async () => queries.getPrices({ from: c }), key),
  );
}

function fetcher(f: () => Promise<any>, key: string) {
  return async () => {
    const start = performance.now();
    const r = await f();
    const end = performance.now();
    console.log(`${key}: ${end - start}ms`);
    return r;
  };
}
