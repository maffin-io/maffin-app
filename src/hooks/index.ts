export { default as useDataSource, DataSourceContext } from './useDataSource';
export type { DataSourceContextType } from './useDataSource';
export { default as useBookStorage } from './useBookStorage';
export { default as useGapiClient } from './useGapiClient';
