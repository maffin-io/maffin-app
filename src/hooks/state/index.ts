export { useTheme } from './useTheme';
export { useOnline } from './useOnline';
