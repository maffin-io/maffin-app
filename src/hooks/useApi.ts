import useSWRImmutable from 'swr/immutable';
import { Fetcher, SWRResponse, SWRConfiguration } from 'swr';

import { Commodity } from '@/book/entities';
import * as queries from '@/lib/queries';
import { PriceDB } from '@/book/prices';
import getUser from '@/lib/getUser';

/**
 * Wrapper around SWR to make the usable keys explicit and re-usable
 */
export default function useApi(key: ApiPaths | null): SWRResponse {
  const f = key !== null ? API[key].f : () => {};
  const swrArgs = key !== null ? API[key].swrArgs : undefined;

  return useSWRImmutable(key, f, swrArgs);
}

export type ApiPaths = (
  '/api/start-date'
  | '/api/main-currency'
  | '/api/commodities'
  | '/api/accounts'
  | '/api/investments'
  | '/api/prices/today'
  | '/api/txs/latest'
  | '/api/user'
);

const API: {
  [key in ApiPaths]: {
    f: Fetcher,
    swrArgs?: SWRConfiguration,
  };
} = {
  '/api/start-date': {
    f: queries.getEarliestDate,
  },
  '/api/main-currency': {
    f: queries.getMainCurrency,
  },
  '/api/commodities': {
    f: Commodity.find,
  },
  '/api/accounts': {
    f: async (k: string) => {
      const start = performance.now();
      const r = await queries.getAccounts();
      const end = performance.now();
      console.log(`${k}: ${end - start}ms`);
      return r;
    },
  },
  '/api/investments': {
    f: queries.getInvestments,
  },
  '/api/prices/today': {
    f: PriceDB.getTodayQuotes,
  },
  '/api/txs/latest': {
    f: queries.getLatestTxs,
  },
  '/api/user': {
    f: getUser,
    swrArgs: {
      refreshInterval: 100000,
      revalidateOnMount: true,
    },
  },
};
