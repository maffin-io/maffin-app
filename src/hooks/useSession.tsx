import React from 'react';
import { useAuth0 } from '@auth0/auth0-react';
import type { User, Auth0ContextInterface } from '@auth0/auth0-react';

import { isStaging } from '@/helpers/env';

const emptyUser: User = {
  name: '',
  email: '',
  picture: '',
};

export type SessionReturn = {
  accessToken: string,
} & Auth0ContextInterface<User>;

/**
 * This hook captures Authorization data from Google Oauth2
 */
export default function useSession(): SessionReturn {
  const auth0 = useAuth0();
  const [accessToken, setAccessToken] = React.useState('');

  /**
   * Note that we have the accessToken attribute
   * in the user profile because we have a custom action
   * in Auth0 that adds it from user_metadata.
   */
  React.useEffect(() => {
    async function load(u: User) {
      setAccessToken(u.accessToken);
    }

    if (auth0.user && auth0.isAuthenticated) {
      load(auth0.user);
    }
  }, [auth0.user, auth0.isAuthenticated, setAccessToken]);

  return {
    ...auth0,
    accessToken,
    isAuthenticated: isStaging() ? true : auth0.isAuthenticated,
    isLoading: isStaging() ? false : auth0.isLoading,
    user: isStaging()
      ? { name: 'Maffin', email: 'iomaffin@gmail.com', picture: '' }
      : auth0.user || emptyUser,
  };
}
