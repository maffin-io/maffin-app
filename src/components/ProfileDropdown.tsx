import Link from 'next/link';
import Image from 'next/image';
import React from 'react';
import { BiLogOut, BiImport } from 'react-icons/bi';

import type BookStorage from '@/apis/BookStorage';
import { useUser, useBookStorage } from '@/hooks';

export default function ProfileDropdown(): JSX.Element {
  const [bookStorage] = useBookStorage();
  const { user } = useUser();

  const fileImportInput = React.useRef<HTMLInputElement>(null);

  return (
    <div className="group relative h-full pl-3 pr-4 rounded-t-sm hover:bg-gunmetal-800">
      <button
        type="button"
        className="flex h-full w-full pl-3 pr-4 items-center"
      >
        {
          user && user.image !== '' && (
            <Image
              src={user.image}
              width={24}
              height={24}
              referrerPolicy="no-referrer"
              className="rounded-full mx-2 w-8 h-8"
              alt="user"
            />
          )
        }
        <div>
          <p className="font-semibold capitalize">
            {user?.name || '...'}
          </p>
          <p className="text-xs">
            {user?.email || ''}
          </p>
        </div>
      </button>
      <ul className="absolute w-40 hidden bg-gunmetal-700 rounded-b-sm py-1 group-hover:block">
        <li className="px-3 py-2 text-sm hover:bg-gunmetal-800">
          <button
            id="menu-item-0"
            type="button"
            role="menuitem"
            tabIndex={-1}
            className="link inline-block w-full whitespace-nowrap"
            onClick={() => fileImportInput.current !== null && fileImportInput.current.click()}
          >
            <BiImport className="inline-block align-middle mr-1" />
            <span className="inline-block align-middle">Import</span>
          </button>
          <input
            type="file"
            ref={fileImportInput}
            onChange={(e) => importBook(e, bookStorage)}
            style={{ display: 'none' }}
          />
        </li>
        <li className="block px-3 py-2 text-sm hover:bg-gunmetal-800">
          <Link
            id="menu-item-1"
            role="menuitem"
            tabIndex={-1}
            href="/user/logout"
            className="inline-block w-full whitespace-nowrap"
          >
            <BiLogOut className="inline-block align-middle mr-1" />
            <span className="inline-block align-middle">Logout</span>
          </Link>
        </li>
      </ul>
    </div>
  );
}

async function importBook(
  event: React.ChangeEvent<HTMLInputElement>,
  bookStorage: BookStorage | null,
) {
  if (event.target.files !== null && event.target.files[0] !== null) {
    const fileReader = new FileReader();
    fileReader.onload = async (loadEvent) => {
      if (loadEvent.target !== null && loadEvent.target.result !== null) {
        if (bookStorage === null) {
          throw new Error('Failed saving, bookStorage is null');
        }
        const rawBook = new Uint8Array(loadEvent.target.result as ArrayBuffer);
        await bookStorage.save(rawBook);
        // TODO: need to re-render the page and reload datasource!
      }
    };

    fileReader.readAsArrayBuffer(event.target.files[0]);
  }
}
