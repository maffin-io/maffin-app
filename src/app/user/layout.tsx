'use client';

import React from 'react';
import Image from 'next/image';

import maffinLogo from '@/assets/images/maffin_logo_sm.png';
import { useTheme } from '@/hooks/state';

export default function UserLayout({
  children,
}: {
  children: React.ReactNode
}): JSX.Element {
  const { data: theme } = useTheme();

  React.useLayoutEffect(() => {
    if (theme === 'dark') {
      document.documentElement.classList.add('dark');
    } else {
      document.documentElement.classList.remove('dark');
    }
  });

  return (
    <div className="flex h-screen place-content-center place-items-center">
      <div className="card flex items-center">
        <Image className="m-0" src={maffinLogo} alt="" height="65" />
        <div className="mt-3 px-4">{children}</div>
      </div>
    </div>
  );
}
