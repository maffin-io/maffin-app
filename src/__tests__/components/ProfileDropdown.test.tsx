import React from 'react';
import { render } from '@testing-library/react';

import ProfileDropdown from '@/components/ProfileDropdown';
import * as sessionHook from '@/hooks/useSession';
import { User } from '@/types/user';

jest.mock('@/hooks/useSession', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/useSession'),
}));

jest.mock('@/components/buttons/ImportButton', () => jest.fn(
  () => <div data-testid="ImportButton" />,
));

describe('ProfileDropdown', () => {
  beforeEach(() => {
    jest.spyOn(sessionHook, 'default').mockReturnValue({
      user: {
        name: '',
        image: '',
        email: '',
      } as User,
    } as sessionHook.SessionReturn);
  });

  it('renders with empty user', () => {
    const { container } = render(<ProfileDropdown />);
    expect(container).toMatchSnapshot();
  });

  it('renders with user', () => {
    jest.spyOn(sessionHook, 'default').mockReturnValue({
      user: {
        name: 'Maffin IO',
        image: 'https://example.com',
        email: 'iomaffin@gmail.com',
      } as User,
    } as sessionHook.SessionReturn);
    const { container } = render(<ProfileDropdown />);
    expect(container).toMatchSnapshot();
  });
});
