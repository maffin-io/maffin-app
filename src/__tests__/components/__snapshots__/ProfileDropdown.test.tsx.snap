// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileDropdown renders with empty user 1`] = `
<div>
  <div
    class="group relative h-full bg-light-100/60 hover:bg-light-100 dark:bg-dark-800/50 dark:hover:bg-dark-800"
  >
    <button
      class="flex h-full w-full pl-3 pr-4 items-center"
      type="button"
    >
      <div>
        <p
          class="fontsemibold capitalize"
        >
          ...
        </p>
        <p
          class="text-xs"
        />
      </div>
    </button>
    <ul
      class="absolute rounded-md w-40 hidden py-1 group-hover:block bg-white dark:bg-dark-700 "
    >
      <li
        class="text-sm hover:bg-light-100 dark:hover:bg-dark-800"
      >
        <div
          data-testid="ImportButton"
        />
      </li>
      <li
        class="text-sm hover:bg-light-100 dark:hover:bg-dark-800"
      >
        <div
          data-testid="ExportButton"
        />
      </li>
      <li
        class="block px-3 py-2 text-sm hover:bg-light-100 dark:hover:bg-dark-800"
      >
        <a
          class="inline-block w-full whitespace-nowrap"
          href="/user/logout"
          id="menu-item-1"
          role="menuitem"
          tabindex="-1"
        >
          <svg
            class="inline-block align-middle mr-1"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M16 13v-2H7V8l-5 4 5 4v-3z"
            />
            <path
              d="M20 3h-9c-1.103 0-2 .897-2 2v4h2V5h9v14h-9v-4H9v4c0 1.103.897 2 2 2h9c1.103 0 2-.897 2-2V5c0-1.103-.897-2-2-2z"
            />
          </svg>
          <span
            class="inline-block align-middle"
          >
            Logout
          </span>
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`ProfileDropdown renders with user 1`] = `
<div>
  <div
    class="group relative h-full bg-light-100/60 hover:bg-light-100 dark:bg-dark-800/50 dark:hover:bg-dark-800"
  >
    <button
      class="flex h-full w-full pl-3 pr-4 items-center"
      type="button"
    >
      <img
        alt="user"
        class="rounded-full mx-2 w-8 h-8"
        data-nimg="1"
        decoding="async"
        height="24"
        loading="lazy"
        referrerpolicy="no-referrer"
        src="/_next/image?url=https%3A%2F%2Fexample.com&w=48&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fexample.com&w=32&q=75 1x, /_next/image?url=https%3A%2F%2Fexample.com&w=48&q=75 2x"
        style="color: transparent;"
        width="24"
      />
      <div>
        <p
          class="fontsemibold capitalize"
        >
          Maffin IO
        </p>
        <p
          class="text-xs"
        >
          iomaffin@gmail.com
        </p>
      </div>
    </button>
    <ul
      class="absolute rounded-md w-40 hidden py-1 group-hover:block bg-white dark:bg-dark-700 "
    >
      <li
        class="text-sm hover:bg-light-100 dark:hover:bg-dark-800"
      >
        <div
          data-testid="ImportButton"
        />
      </li>
      <li
        class="text-sm hover:bg-light-100 dark:hover:bg-dark-800"
      >
        <div
          data-testid="ExportButton"
        />
      </li>
      <li
        class="block px-3 py-2 text-sm hover:bg-light-100 dark:hover:bg-dark-800"
      >
        <a
          class="inline-block w-full whitespace-nowrap"
          href="/user/logout"
          id="menu-item-1"
          role="menuitem"
          tabindex="-1"
        >
          <svg
            class="inline-block align-middle mr-1"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M16 13v-2H7V8l-5 4 5 4v-3z"
            />
            <path
              d="M20 3h-9c-1.103 0-2 .897-2 2v4h2V5h9v14h-9v-4H9v4c0 1.103.897 2 2 2h9c1.103 0 2-.897 2-2V5c0-1.103-.897-2-2-2z"
            />
          </svg>
          <span
            class="inline-block align-middle"
          >
            Logout
          </span>
        </a>
      </li>
    </ul>
  </div>
</div>
`;
